### PR TITLE
fix(Core/Loot): Items below group threshold should not be blocked by master looter.

### LINKED ISSUE #4939: Equipment picking mode bug [$10]
##### CURRENT BEHAVIOUR:
When `loot:master looter` mode is set, no matter what items are allocated by the `master looter`, even if you set epic quality

##### EXPECTED BLIZZLIKE BEHAVIOUR:
When `loot:master looter` mode is set, and then choose `epic`,then all items below `epic` quality will be picked up by all players in the team in turn

##### STEPS TO REPRODUCE THE PROBLEM:
1. set `loot:master looter` mode

2. set `epic` 

3. Then set up a monster to drop `gray` items

##### AC HASH/COMMIT:
https://github.com/azerothcore/azerothcore-wotlk/commit/cb0fb719db050b2c6b7651c5ade044b3d318c453

##### OPERATING SYSTEM:
ubuntu 18.04

##### MODULES:
none

<bountysource-plugin>

---
There is a **[$10 open bounty](https://www.bountysource.com/issues/97335622-equipment-picking-mode-bug?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by UltraNix**
## Changes Proposed:
-  Added additional check if items is under group threshold.
-  

## Issues Addressed:
- Closes #4939.

## Tests Performed:

- Tested ingame by 2 chars in Utgarde Pinnacle.
- 

## How to Test the Changes:

1. Make a party
2. Teleport both players to any dungeon, e.g. Utgarde Pinnacle
3. Set master loot and item threshold to epic
4. Kill random mobs and try to loot gray/white items by non-master looter.

## Known Issues and TODO List:

- [ ]
- [ ] 

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by UltraNix** [comment]
@mpfans Please, test it.

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by mpfans** [comment]
under the team leader allocation mode, the allocation modes of purple, blue and green were tested. it‘s work! good job!

**Comment by mpfans** [comment]
At present, there is still a bug, that is, in the  `loot:master looter` mode, the team leader cannot allocate professional formula to other team-mates

**Comment by UltraNix** [comment]
You mean recipes?

**Comment by mpfans** [comment]
> You mean recipes?

if `loot:master looter` mode ,  boss loot a item (id = 32753),if master not learn `Tailoring` professional skills,then he can't see the item, and he can't assign a picker to the item

**Comment by pklloveyou** [comment]
TestedI tested all the loot patterns
image
The repair is effective
In addition, there are also problems in the demand model. Hooligans and shamans can demand lock armor at the same time

File: src/server/game/Loot/LootMgr.cpp [hunk 1]
```diff
@@ -397,17 +397,31 @@ bool LootItem::AllowedForPlayer(Player const* player) const
...
     if ((pProto->Flags2 & ITEM_FLAGS_EXTRA_HORDE_ONLY) && player->GetTeamId(true) != TEAM_HORDE)
         return false;
 
     if ((pProto->Flags2 & ITEM_FLAGS_EXTRA_ALLIANCE_ONLY) && player->GetTeamId(true) != TEAM_ALLIANCE)
         return false;
 
+    // Master looter can see certain items even if the character can't loot them
+    if (!isGivenByMasterLooter && player->GetGroup() && player->GetGroup()->GetMasterLooterGuid() == player->GetGUID())
+    {
+        // check quest requirements
```
Comment (mpfans) [inline src/server/game/Loot/LootMgr.cpp:410]: Is it a repetition of the following request requirements?

Comment (UltraNix) [reply src/server/game/Loot/LootMgr.cpp:410]: Yep, but only for master looter.

**Comment by mpfans** [comment]
@UltraNix

**Comment by mpfans** [comment]
> TestedI tested all the loot patterns
> image
> The repair is effective
> In addition, there are also problems in the demand model. Hooligans and shamans can demand lock armor at the same time

This is tradable, so there is no bug

**Comment by UltraNix** [comment]
@mpfans That's because there is a condition for this item that checks whether you have BS skill. On TC it's the same - I'm not sure, if this is Blizzlike.

**Comment by mpfans** [comment]
@UltraNix like Blizzard will check if it has learned the relevant skills. However, in the`master looter` mode, the master can see this item. If master has not learned the relevant skills, he cannot pick it up, but he can choose to assign it to the players who have learned the skills.

**Comment by mpfans** [comment]
So if the master cannot see the item, no one can pick it up

**Comment by UltraNix** [comment]
@mpfans Try again.

**Comment by mpfans** [comment]
> @mpfans Try again.

I'll give you feedback in about eight hours

File: src/server/game/Loot/LootMgr.cpp [hunk 2]
```diff
@@ -408,6 +417,24 @@ bool LootItem::AllowedForPlayer(Player const* player) const
...
         return false;
 
+    // Master looter can see certain items even if the character can't loot them
+    if (!isGivenByMasterLooter && isMasterLooter)
+    {
+        // check quest requirements
+        if (!(pProto->FlagsCu & ITEM_FLAGS_CU_IGNORE_QUEST_STATUS) && ((needs_quest || (pProto->StartQuest && player->GetQuestStatus(pProto->StartQuest) != QUEST_STATUS_NONE)) && !player->HasQuestForItem(itemid)))
+            return false;
+
+        return true;
```
Comment (Helias) [inline src/server/game/Loot/LootMgr.cpp:427]:
```suggestion
        return !(!(pProto->FlagsCu & ITEM_FLAGS_CU_IGNORE_QUEST_STATUS) && ((needs_quest || (pProto->StartQuest && player->GetQuestStatus(pProto->StartQuest) != QUEST_STATUS_NONE)) && !player->HasQuestForItem(itemid)))
```

**Comment by mpfans** [comment]
> @mpfans Try again.
Everything worked,but another bug is that the master can't see the quest items that have completed the quest, and can't assign them to the players who haven't completed the quest.

**Comment by UltraNix** [comment]
@mpfans In that case i believe that items are not green/blue/epic, so player can take it anyway?

**Comment by mpfans** [comment]
> @mpfans In that case i believe that items are not green/blue/epic, so player can take it anyway?

no,have epic item (id 32405).. but if the master has just completed the quest, he will not see the quest item..

**Comment by UltraNix** [comment]
Okay.

File: src/server/game/Loot/LootMgr.cpp (hunk 2 above)
Comment (UltraNix) [reply src/server/game/Loot/LootMgr.cpp:427]: For readability reason, I think it's okay now.

**Comment by mpfans** [comment]
> Okay.

All items work except the completed quest items that cannot be seen and assigned

**Comment by UltraNix** [comment]
Check again.

**Comment by mpfans** [comment]
> Check again.

ok,Please wait for my feedback

**Comment by UltraNix** [comment]
@mpfans ?

File: src/server/game/Loot/LootMgr.cpp (hunk 2 above)
Comment (Helias) [reply src/server/game/Loot/LootMgr.cpp]: ok :+1:

**Comment by mpfans** [comment]
> @mpfans ?

Sorry, I'm on vacation these days,after testing, the master  completed quest still can't see the task items (id 32405)

**Comment by UltraNix** [comment]
@mpfans Please, once again ;)

**Comment by mpfans** [comment]
good job!!! no more questions

**Comment by FrancescoBorzi** [comment]
good job @UltraNix and thanks @pklloveyou and @mpfans for your precious feedback <3

**Comment by mpfans** [comment]
Our test team found a new bug. In `Stratholme` instance, no one in the team has a quest (Quest ID 5212), but when the master loot mode, the monster will drop (item 13174) and no one can pick it up,if it is an animal, the player will not be able to skinning

**Comment by mpfans** [comment]
> Our test team found a new bug. In `Stratholme` instance, no one in the team has a quest (Quest ID 5212), but when the master loot mode, the monster will drop (item 13174) and no one can pick it up,if it is an animal, the player will not be able to skinning

only master loot mode can see it,quest items are not visible in other modes

**Comment by UltraNix** [comment]
Yep - this is the bug related to recent fix. Either master looter will not see such quest items (no matter if already completed quest), or the quest items will be visible always to master looter.

**Comment by mpfans** [comment]
> Yep - this is the bug related to recent fix. Either master looter will not see such quest items (no matter if already completed quest), or the quest items will be visible always to master looter.

So it can't be fixed?

**Comment by UltraNix** [comment]
Hmm.
But you said none of you had the quest and the quest item dropped anyway? It's something wrong here.

**Comment by mpfans** [comment]
> Hmm.
> But you said none of you had the quest and the quest item dropped anyway? It's something wrong here.

Yes, no one has geted a quest. Killing a quest NPC will also can see quest items. But only the master mode can see it. Other modes cannot be seen.

**Comment by UltraNix** [comment]
@mpfans "Test it baby one more time..."

**Comment by pklloveyou** [comment]
image
Tested

**Comment by mpfans** [comment]
1.In the master looter mode, the master can be seen and assigned professional recipe to the player (work)
2.In the master looter mode, the master can be seen and assigned quest item to the player (work)
3.In the master looter mode, the master can distribute items to players according to the set quality (work)

so there are no bugs

**Pull request merged**